### PR TITLE
internal/v5: new API version (currently identical to v4)

### PR DESCRIPTION
**Pull request by rogpeppe**
We have made no semantic changes so far. We factor out ResolveURL so that
it can be changed independently in an API version. We duplicate all
the tests so that even though v4 is using the v5 code, we're testing
the whole thing. Some tests were removed from the v4 fork because
they were testing internal-only functionality.

These were:

```
APISuite.TestMetaPermAudit
ArchiveSuite.TestBundleCharms
ArchiveSuite.TestGetNewPromulgatedRevision
authSuite.TestGroupsForUserSuccess
authSuite.TestGroupsForUserWithNoIdentity
authSuite.TestGroupsForUserWithInvalidIdentityURL
authSuite.TestGroupsForUserWithInvalidBody
authSuite.TestGroupsForUserWithErrorResponse
authSuite.TestGroupsForUserWithBadErrorResponse
APISuite.TestProcessIconWorksOnDefaultIcon
APISuite.TestProcessIconDoesNotQuoteNewlines
SearchSuite.TestParseSearchParams
```

and all the audit tests (we are assuming that because v4 does not interfere with any
of the calls themselves, it is not possible for it to avoid the usual auditing).

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1351/
Test FAILed.

File: internal/v4/archive_test.go [hunk 1]
```diff
@@ -1707,12 +1522,12 @@ func patchArchiveCacheAges(s interface {
 // It assumes that patchArchiveCacheAges has been called
 // for the current test.
 func assertCacheControl(c *gc.C, h http.Header, isPublic bool) {
-	if isPublic {
-		seconds := 5
-		c.Assert(h.Get("Cache-Control"), gc.Equals, fmt.Sprintf("public, max-age=%d", seconds))
-	} else {
-		c.Assert(h.Get("Cache-Control"), gc.Equals, "no-cache, must-revalidate")
-	}
+	//	if isPublic {
```
Comment (mhilton) [inline internal/v4/archive_test.go]: Do something with this, don't leave it commented out.

**Comment by mhilton** [comment]
:+1: with a small nit

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1352/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1353/
Test PASSed.

File: server.go [hunk 2]
```diff
@@ -17,16 +17,19 @@ import (
 	"gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/legacy"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/v4"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/v5"
 )
 
 // Versions of the API that can be served.
 const (
 	V4     = "v4"
+	V5     = "v5"
```
Comment (urosj) [inline server.go:27]: Purely pedantic remark, can we keep descending order here? Or ascending. Just not mixed :)

File: server.go [hunk 3]
```diff
@@ -17,16 +17,19 @@ import (
...
 	Legacy = ""
 )
 
 var versions = map[string]charmstore.NewAPIHandlerFunc{
 	V4:     v4.NewAPIHandler,
+	V5:     v5.NewAPIHandler,
```
Comment (urosj) [inline server.go:33]: Same bike shedding as above :)

**Comment by urosj** [comment]
:+1: with a small remark, no big PR should pass without a bit of bike shedding :)

File: internal/v4/archive_test.go (hunk 1 above)
Comment (rogpeppe) [inline internal/v4/archive_test.go]: done. exported the relevant constant and reenabled the checks in v4.

File: server.go (hunk 2 above)
Comment (rogpeppe) [inline server.go:27]: done

File: server.go (hunk 3 above)
Comment (rogpeppe) [inline server.go:33]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1354/
Test PASSed.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1355/
Test PASSed.

**Pull request merged**